### PR TITLE
[Internal] Build report internal plugin

### PR DESCRIPTION
**Pull request by yoannmoinet**
### What and why?

We need a cross-bundler report of any build.
It should act as a universal layer between the bundler and our plugins.
This way, any other internal plugin can leverage the report (wink wink, the telemetry plugin).

### How?

Introducing the new internal `build-report`.
Mostly based on the output of each bundler we support (esbuild, webpack4/5, rollup, vite).
Minus some particularities depending on what the bundler actually shares.

We end up with a report of the form:

```javascript
{
  errors: [], // Array of errors,
  warnings: [], // Array of warnings,
  start: 1234567890, // Timestamp of beginning of the build.
  inputs: [ // List of all the modules used in the build.
    {
      name: 'moduleName',
      filepath: 'absolute/path',
      dependencies: [], // List of the dependencies of this module, as inputs.
      dependents: [],  // List of the dependents of this module, as inputs.
      size: 1234,
      type: 'js',
    },
    {...},
  ],
  outputs: [ // List of all the assets produced by the build.
    {
      name: 'assetName',
      filepath: 'absolute/path',
      inputs: [], // List of inputs included in this asset.
      size: 12345,
      type: 'js',
    },
    {...},
  ],
  entries: [
    {
      name: 'entryName',
      inputs: [], // List of inputs included in this entry.
      outputs: [], // List of outputs included in this entry.
      size: 123456,
      type: 'js',
    },
    {...},
  ]
}
```

This will be used in https://github.com/DataDog/build-plugins/pull/93 to remove any legacy code that was specific to webpack and esbuild.
Producing the same metrics for every bundler.

> [!NOTE]
> Tests are failing tests, because some were over-written by the new telemetry plugin. And splitting the PRs was complicated. Waiting for https://github.com/DataDog/build-plugins/pull/93 to be merged into this one to make it pass.

### CODE REVIEW

File: packages/tests/src/core/plugins/build-report/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    return {
+        ...originalModule,
+        sendMetrics: jest.fn(() => []),
+    };
+});
+
+const outputTextsMocked = jest.mocked(outputTexts);
+
+const sortFiles = (a: File | Output | Entry, b: File | Output | Entry) => {
+    if (a.name < b.name) {
```
Comment (elbywan) [inline packages/tests/src/core/plugins/build-report/index.test.ts:44]: using [String.prototype.localeCompare()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) could simplify this maybe?

```ts
// something like:
return a.name.localeCompare(b.name)
```

**Review by elbywan — APPROVED** [review]
LGTM with some comments 👍

File: packages/core/src/plugins/build-report/esbuild.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+            const cwd = context.cwd;
+
+            // Store entry names based on the configuration.
+            const entrypoints = build.initialOptions.entryPoints;
+            const entryNames = new Map();
+            if (Array.isArray(entrypoints)) {
+                // We don't have an indexed object as entry, so we can't get an entry name from it.
+                for (const entry of entrypoints) {
+                    const fullPath = entry && typeof entry === 'object' ? entry.in : entry;
+                    const realEntry = cleanName(context, fullPath);
```
[the remark this answers is not in the document]
Comment (yoannmoinet) [reply packages/core/src/plugins/build-report/esbuild.ts:37]: Oh, good call.
I'll have to handle this case too 😓

File: packages/core/src/plugins/build-report/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        ...buildReport,
+        entries,
+        inputs,
+        outputs,
+    };
+};
+
+// Make list of paths unique, remove the current file and particularities.
+export const cleanReport = (report: string[], filepath: string) => {
+    const particularities = ['unknown', 'commonjsHelpers.js', 'vite/preload-helper.js'];
```
[the remark this answers is not in the document]
Comment (yoannmoinet) [reply packages/core/src/plugins/build-report/helpers.ts:179]: Good call!

File: packages/core/src/plugins/build-report/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        (reportFilepath) =>
+            reportFilepath !== filepath && !particularities.includes(reportFilepath),
+    );
+};
+
+// Clean a path from its query parameters and leading invisible characters.
+export const cleanPath = (filepath: string) => {
+    let resolvedPath = filepath;
+    try {
+        resolvedPath = require.resolve(filepath);
```
[the remark this answers is not in the document]
Comment (yoannmoinet) [reply packages/core/src/plugins/build-report/helpers.ts:191]: 🤔 wondering the same.
It's probably a remain of my iterated implementation when I was dealing with unresolved paths like `@dd/something`. 
But I'm not using the same sources of information anymore.
So it doesn't seem necessary anymore.

I'll try and remove it, see if the tests are still running.

File: packages/tests/src/core/plugins/build-report/index.test.ts (hunk 1 above)
Comment (yoannmoinet) [reply packages/tests/src/core/plugins/build-report/index.test.ts:44]: Amazing 🤯

**Review by Miz85 — APPROVED** [review]

File: packages/core/src/plugins/build-report/helpers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yoannmoinet) [reply packages/core/src/plugins/build-report/helpers.ts]: So this is still required, but I've reduced its usage to entry file path only.

**Pull request merged**